docs/source/daml/reference/contract-keys.rst

```diff
@@ -0,0 +1,95 @@
+   :start-after: -- start contract key setup snippet
+   :end-before: -- end contract key setup snippet
+
+What can be a contract key
+**************************
+
+The key should either be a tuple or a record, and it **must** include every party that you specify as a ``maintainer`` (see `Specifying maintainers`_ below).
+
+For example, with ``maintainer bank``, ``key (bank, number) : (Party, Text)`` is valid; but wouldn't be if you removed ``bank`` from the key.
+
+It's best to use simple types for your keys like ``Text`` or ``Integer``, rather than a list or more complex type.
+
+Specifying maintainers
+**********************
+
+If you specify a contract key for a template, you must also specify a ``maintainer`` or maintainers, in a similar way to specifying signatories or observers. Maintainers are the parties that know about all of the keys that they are party to, and are used by the engine to guarantee uniqueness of contract keys.  The maintainers **must** be signatories or observers of the contract.
+
+Keys are unique only to their maintainers. For example, say you have a key that you're using as the identifer for a ``BankAccount`` contract. You might have ``key (bank, accountId) : (Party, Text)``. When you create a new bank account, the contract key ensures that no-one else can have an account with the same ``accountID`` at that bank. But that doesn't apply to other banks: for a contract with a different bank as maintainer, you could happily re-use that ``accountID``.
+
+When you're writing DAML models, the maintainers only matter since they affect authorization -- much like signatories and observers. You don't need to do anything to "maintain" the keys.
+
+Checking of the keys is done automatically at execution time, by the DAML exeuction engine: if someone tries to create a new contract that duplicates an existing contract key, the execution engine will cause that creation to fail. 
+
+Contract keys functions
+***********************
+
+Contract keys introduce several new functions.
+
+``fetchByKey``
+==============
+
+``fetchedContract <- fetchByKey @ContractType contractKey``
+
+Use ``fetchByKey`` to fetch the ID and data of the contract with the specified key. It is an alternative to the currently-used ``fetch``.
+
+It returns a tuple of the ID and the contract object (containing all its data). 
+
+You need authorization from **at least one** of the maintainers to run ``fetchByKey``. A maintainer can authorize by being a signatory, or by submitting the command/being a controller for the choice.
+
+``fetchByKey`` fails and aborts the transaction if:
+
+- you don't have sufficient authorization
+- you're not a stakeholder of the contract you're trying to fetch
+
+This means that if it fails, it doesn't guarantee that a contract with that key doesn't exist, just that you can't see one.
+
+Because the type is ambiguous, when calling you need to specify what you expect with ``@`` and the type, for example ``@MyTemplateType``.
```

Comment:
```suggestion
Because different templates can use the same key type, you need to specify the type of the contract you are trying to fetch using the ``@ContractType`` syntax.
```
Same below for `lookupByKey` and `exerciseByKey`.